### PR TITLE
feat(maitake): add an async `Mutex`

### PR DESCRIPTION
**Pull request by hawkw**
This branch adds an async `Mutex` type to `maitake`, implemented using
`WaitQueue`. The implementation should be pretty straightforward,
although it required adding some additional methods to `WaitQueue`.
Currently, all the new `WaitQueue` functions are `pub(crate)`, but I'm
open to making them public APIs as well; I just wanted to save that for
a separate PR.

The `Mutex` type lives in a new `maitake::sync` module for
synchronization primitives --- I was on the fence about whether it still
makes sense to have a separate `maitake::wait` module if we are going to
have `maitake::sync`, but I didn't want to put the mutex in `wait`, as
it does more than *just* notify waiting tasks. Perhaps the `wait` types
should be moved into `sync`, but I'd prefer to do that in a separate PR.

### CODE REVIEW

File: maitake/src/sync/mutex.rs [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+///
+/// [^1]: And therefore require an operating system to manage threading.
+///
+/// [^2]: The [tasks][crate::task::Task] themselves must, of course, be stored
+///     somewhere, but this need not be a heap allocation in systems with a
+///     fixed set of  statically-allocated tasks. And, when tasks *are*
+///     heap-allocated, these  allocations [need not be provided by
+///     `liballoc`][storage].
+///
+/// [^3]: In fact, this mutex _cannot_ implement poisoning, as poisoning
```
Comment (jamesmunns) [inline maitake/src/sync/mutex.rs:62]: Is this documented anywhere else?

File: maitake/src/sync/mutex.rs [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+            _wake: WakeOnDrop(self),
+            data: self.data.get_mut(),
+        }
+    }
+}
+
+impl<T: ?Sized + fmt::Debug> fmt::Debug for Mutex<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Mutex")
+            .field("data", &fmt::opt(&self.try_lock()).or_else("<locked>"))
```
Comment (jamesmunns) [inline maitake/src/sync/mutex.rs:238]: :100:

**Review by jamesmunns — APPROVED** [review]
Overall lgtm!

File: maitake/src/sync/mutex.rs (hunk 1 above)
Comment (hawkw) [reply maitake/src/sync/mutex.rs:62]: no, i keep meaning to add it to the crate level docs

Comment (jamesmunns) [reply maitake/src/sync/mutex.rs:62]: Not a blocker! I just wanted to cross link it if there was a larger discussion somewhere.

**Pull request merged**